packages/next/src/build/segment-config/app/app-segment-config.ts

```diff
@@ -1,6 +1,95 @@
 import { z } from 'next/dist/compiled/zod'
 import { formatZodError } from '../../../shared/lib/zod'
 
+const CookieSchema = z
+  .object({
+    name: z.string(),
+    value: z.string(),
+    httpOnly: z.boolean().optional(),
+    path: z.string().optional(),
+  })
+  .strict()
+
+const RuntimeSampleSchema = z
+  .object({
+    cookies: z.array(CookieSchema).optional(),
+    headers: z.array(z.tuple([z.string(), z.string()])).optional(),
+    params: z.record(z.union([z.string(), z.array(z.string())])).optional(),
+    searchParams: z
+      .record(z.union([z.string(), z.array(z.string()), z.undefined()]))
+      .optional(),
+  })
+  .strict()
+
+const StaticPrefetchSchema = z
+  .object({
+    mode: z.literal('static'),
+    from: z.array(z.string()).optional(),
+    expectUnableToVerify: z.boolean().optional(),
+  })
+  .strict()
+
+const RuntimePrefetchSchema = z
+  .object({
+    mode: z.literal('runtime'),
+    samples: z.array(RuntimeSampleSchema).min(1),
+    from: z.array(z.string()).optional(),
+    expectUnableToVerify: z.boolean().optional(),
+  })
+  .strict()
+
+const PrefetchSchema = z.discriminatedUnion('mode', [
+  StaticPrefetchSchema,
+  RuntimePrefetchSchema,
+])
+
+export type Prefetch = __GenericPrefetch | StaticPrefetch | RuntimePrefetch
+// the __GenericPrefetch type is used to avoid type widening issues with
+// our choice to make exports the medium for programming a Next.js application
+// With exports the type is controlled by the module and all we can do is assert on it
+// from a consumer. However with string literals in objects these are by default typed widely
+// and thus cannot match the discriminated union type. If we figure out a better way we should
+// delete the __GenericPrefetch member.
+interface __GenericPrefetch {
+  mode: string
+  samples?: Array<WideRuntimeSample>
+  from?: string[]
+  expectUnableToVerify?: boolean
+}
+interface StaticPrefetch {
+  mode: 'static'
+  from?: string[]
+  expectUnableToVerify?: boolean
+}
+interface RuntimePrefetch {
+  mode: 'runtime'
+  samples: Array<RuntimeSample>
+  from?: string[]
+  expectUnableToVerify?: boolean
+}
+type WideRuntimeSample = {
+  cookies?: Array<{
+    name: string
+    value: string
+    httpOnly?: boolean
+    path?: string
+  }>
+  headers?: Array<string[]>
+  params?: { [key: string]: string | string[] }
+  searchParams?: { [key: string]: string | string[] | undefined }
+}
```

Comment:
i'd prefer to restructure this to not duplicate everything between `RuntimeSample` and `WideRuntimeSample`:
```suggestion
type WideRuntimeSample = {
  cookies?: RuntimeSample['cookies']
  // This is meant to be an array of tuples, but it won't be inferred that way by default
  headers?: Array<string[][]>
  params?: RuntimeSample['params']
  searchParams?: RuntimeSample['searchParams']
}
```
AFAICT `headers` is the only one that needs to differ here, so let's just reference the other fields directly